File: vessel_lists/analytics.py
from fastapi import APIRouter, Depends, Response, Query, HTTPException, Request
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List, Optional
import csv
import io

import models, schemas, database
from security import sanitize_filename, audit_log

router = APIRouter()

def get_db():
    db = database.SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("/export/list/{list_id}", tags=["Analytics"])
def export_list_csv(list_id: int, request: Request, db: Session = Depends(get_db)):
    
    vessel_list = db.query(models.VesselList).filter(models.VesselList.id == list_id).first()
    if not vessel_list:
        raise HTTPException(status_code=404, detail="List not found")

    vessels = db.query(models.Vessel).filter(models.Vessel.list_id == list_id).all()

    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow(['MMSI', 'IMO', 'Name', 'Flag', 'LastPosition', 'Note'])

    for vessel in vessels:
        writer.writerow([
            vessel.mmsi,
            vessel.imo or '',
            vessel.name or '',
            vessel.flag or '',
            vessel.lastposition or '',
            vessel.note or ''
        ])

    csv_data = output.getvalue()
    output.close()

    safe_filename = sanitize_filename(vessel_list.name)

    audit_log("EXPORT", "vessel_list", list_id,
              {"format": "csv", "vessel_count": len(vessels)}, request)

    return Response(
        content=csv_data,
        media_type="text/csv",
        headers={
            "Content-Disposition": f"attachment; filename={safe_filename}.csv"
        }
    )


@router.get("/stats", tags=["Analytics"])
def get_stats(db: Session = Depends(get_db)):
    
    
    total_lists = db.query(func.count(models.VesselList.id)).scalar()
    total_vessels = db.query(func.count(models.Vessel.id)).scalar()
    
    vessels_per_list = db.query(
        models.VesselList.name,
        models.VesselList.color,
        func.count(models.Vessel.id).label('count')
    ).outerjoin(models.Vessel).group_by(models.VesselList.id).all()
    
    flags = db.query(
        models.Vessel.flag,
        func.count(models.Vessel.id).label('count')
    ).filter(models.Vessel.flag.isnot(None)).group_by(models.Vessel.flag).all()
    
    with_imo = db.query(func.count(models.Vessel.id)).filter(models.Vessel.imo.isnot(None)).scalar()
    without_imo = total_vessels - with_imo
    
    with_position = db.query(func.count(models.Vessel.id)).filter(models.Vessel.lastposition.isnot(None)).scalar()
    
    return {
        "overview": {
            "total_lists": total_lists,
            "total_vessels": total_vessels,
            "unique_flags": len(flags),
            "with_imo": with_imo,
            "without_imo": without_imo,
            "with_position": with_position
        },
        "lists": [
            {
                "name": name,
                "color": color,
                "vessel_count": count
            }
            for name, color, count in vessels_per_list
        ],
        "flags": [
            {
                "flag": flag,
                "count": count
            }
            for flag, count in flags
        ]
    }


@router.get("/vessels/advanced-search", tags=["Vessels"])
def advanced_search(
    mmsi: Optional[str] = Query(None, description="Search by MMSI (partial match)"),
    imo: Optional[str] = Query(None, description="Search by IMO (partial match)"),
    name: Optional[str] = Query(None, description="Search by vessel name (partial match)"),
    flag: Optional[str] = Query(None, description="Filter by flag (exact match)"),
    list_id: Optional[int] = Query(None, description="Filter by list ID"),
    has_imo: Optional[bool] = Query(None, description="Filter vessels with/without IMO"),
    has_position: Optional[bool] = Query(None, description="Filter vessels with/without position"),
    db: Session = Depends(get_db)
):
    query = db.query(models.Vessel).join(models.VesselList)
    
    if mmsi:
        query = query.filter(models.Vessel.mmsi.contains(mmsi))
    if imo:
        query = query.filter(models.Vessel.imo.contains(imo))
    if name:
        query = query.filter(models.Vessel.name.contains(name))
    if flag:
        query = query.filter(models.Vessel.flag == flag)
    if list_id:
        query = query.filter(models.Vessel.list_id == list_id)
    if has_imo is not None:
        if has_imo:
            query = query.filter(models.Vessel.imo.isnot(None))
        else:
            query = query.filter(models.Vessel.imo.is_(None))
    if has_position is not None:
        if has_position:
            query = query.filter(models.Vessel.lastposition.isnot(None))
        else:
            query = query.filter(models.Vessel.lastposition.is_(None))
    
    results = query.all()
    
    return [
        {
            "id": vessel.id,
            "mmsi": vessel.mmsi,
            "imo": vessel.imo,
            "name": vessel.name,
            "flag": vessel.flag,
            "lastposition": vessel.lastposition,
            "note": vessel.note,
            "list_id": vessel.list_id,
            "list_name": vessel.vessel_list.name,
            "list_color": vessel.vessel_list.color
        }
        for vessel in results
    ]


@router.get("/filters/flags", tags=["Analytics"])
def get_available_flags(db: Session = Depends(get_db)):
    
    flags = db.query(models.Vessel.flag).filter(
        models.Vessel.flag.isnot(None)
    ).distinct().all()
    return [f[0] for f in flags]

@router.get("/filters/lists", tags=["Analytics"])
def get_lists_summary(db: Session = Depends(get_db)):
    
    lists = db.query(models.VesselList).all()
    return [
        {
            "id": l.id,
            "name": l.name,
            "color": l.color,
            "vessel_count": len(l.vessels)
        }
        for l in lists
    ]


@router.get("/vessels/aggregated", tags=["Analytics"])
def get_aggregated_vessels(db: Session = Depends(get_db)):
    all_vessels = db.query(models.Vessel).join(models.VesselList).all()
    
    vessel_groups = {}
    
    for vessel in all_vessels:
        key = vessel.mmsi if vessel.mmsi else f"imo_{vessel.imo}"
        
        if key not in vessel_groups:
            vessel_groups[key] = {
                "mmsi": vessel.mmsi,
                "imo": vessel.imo,
                "name": vessel.name,
                "flag": vessel.flag,
                "lastposition": vessel.lastposition,
                "note": vessel.note,
                "lists": [],
                "list_count": 0
            }
        
        vessel_groups[key]["lists"].append({
            "list_id": vessel.vessel_list.id,
            "list_name": vessel.vessel_list.name,
            "list_color": vessel.vessel_list.color,
            "vessel_id": vessel.id
        })
        vessel_groups[key]["list_count"] = len(vessel_groups[key]["lists"])
    
    result = sorted(vessel_groups.values(), key=lambda x: x["list_count"], reverse=True)
    
    return {
        "total_unique_vessels": len(result),
        "vessels": result
    }

@router.get("/export/aggregated", tags=["Analytics"])
def export_aggregated_csv(request: Request, db: Session = Depends(get_db)):
    
    data = get_aggregated_vessels(db)

    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow(['MMSI', 'IMO', 'Name', 'Flag', 'Lists', 'List Count', 'List Names'])

    for vessel in data["vessels"]:
        list_names = ', '.join([l['list_name'] for l in vessel['lists']])
        writer.writerow([
            vessel['mmsi'] or '',
            vessel['imo'] or '',
            vessel['name'] or '',
            vessel['flag'] or '',
            vessel['list_count'],
            vessel['list_count'],
            list_names
        ])

    csv_data = output.getvalue()
    output.close()

    audit_log("EXPORT", "aggregated_vessels", None,
              {"format": "csv", "vessel_count": data["total_unique_vessels"]}, request)

    return Response(
        content=csv_data,
        media_type="text/csv",
        headers={
            "Content-Disposition": "attachment; filename=aggregated_vessels.csv"
        }
    )

File: vessel_lists/frontend/src/analytics.js
import { api } from './api.js';

export const analytics = {
    async loadStats() {
        try {
            const stats = await api.getStats();

            document.getElementById('stat-lists').textContent = stats.overview.total_lists;
            document.getElementById('stat-vessels').textContent = stats.overview.total_vessels;
            document.getElementById('stat-flags').textContent = stats.overview.unique_flags;
            document.getElementById('stat-imo').textContent = stats.overview.with_imo;
            document.getElementById('stat-position').textContent = stats.overview.with_position;

            const flagSelect = document.getElementById('filter-flag');
            const listSelect = document.getElementById('filter-list');

            stats.flags.forEach(flagData => {
                const option = document.createElement('option');
                option.value = flagData.flag;
                option.textContent = `${flagData.flag} (${flagData.count})`;
                flagSelect.appendChild(option);
            });

            stats.lists.forEach(list => {
                const option = document.createElement('option');
                option.value = list.name;
                option.textContent = `${list.name} (${list.vessel_count})`;
                option.dataset.listId = list.vessel_count; // Store for later use
                listSelect.appendChild(option);
            });

        } catch (error) {
            console.error('Failed to load stats:', error);
        }
    },

    bindAdvancedSearch() {
        const searchBtn = document.getElementById('advanced-search-btn');
        const clearBtn = document.getElementById('clear-filters-btn');

        searchBtn.addEventListener('click', async () => {
            const filters = {
                mmsi: document.getElementById('filter-mmsi').value,
                imo: document.getElementById('filter-imo').value,
                name: document.getElementById('filter-name').value,
                flag: document.getElementById('filter-flag').value,
                has_imo: document.getElementById('filter-has-imo').value,
            };

            try {
                const results = await api.advancedSearch(filters);
                this.renderSearchResults(results);
                document.getElementById('list-grid').style.display = 'none';
            } catch (error) {
                alert('Search failed: ' + error.message);
            }
        });

        clearBtn.addEventListener('click', () => {
            document.getElementById('filter-mmsi').value = '';
            document.getElementById('filter-imo').value = '';
            document.getElementById('filter-name').value = '';
            document.getElementById('filter-flag').value = '';
            document.getElementById('filter-list').value = '';
            document.getElementById('filter-has-imo').value = '';
            document.getElementById('search-results').innerHTML = '';
            document.getElementById('list-grid').style.display = 'grid';
        });
    },

    renderSearchResults(results) {
        const container = document.getElementById('search-results');

        if (results.length === 0) {
            container.innerHTML = '<p style="opacity: 0.7; text-align: center; padding: 2rem;">No vessels found matching your criteria.</p>';
            return;
        }

        container.innerHTML = `
            <div style="background: var(--card-bg); border: var(--card-border); border-radius: 0.75rem; padding: 1rem; margin-bottom: 1rem;">
                <strong>${results.length} vessel(s) found</strong>
            </div>
        `;

        const resultsDiv = document.createElement('div');
        resultsDiv.className = 'search-results';

        results.forEach(vessel => {
            const item = document.createElement('div');
            item.className = 'search-result-item';
            item.style.borderLeftColor = vessel.list_color;
            item.innerHTML = `
                <div class="search-result-info">
                    <div><strong>MMSI:</strong> ${vessel.mmsi} ${vessel.imo ? `| <strong>IMO:</strong> ${vessel.imo}` : ''}</div>
                    ${vessel.name ? `<div><strong>Name:</strong> ${vessel.name}</div>` : ''}
                    ${vessel.flag ? `<div><strong>Flag:</strong> ${vessel.flag}</div>` : ''}
                    <div style="margin-top: 0.5rem;">
                        <span class="search-result-badge" style="background-color: ${vessel.list_color}33;">
                            <span class="list-badge" style="background-color: ${vessel.list_color}"></span>
                            ${vessel.list_name}
                        </span>
                    </div>
                </div>
            `;
            resultsDiv.appendChild(item);
        });

        container.appendChild(resultsDiv);
    },

    async exportListCSV(listId, listName) {
        try {
            const blob = await api.exportListCSV(listId);
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `${listName.replace(/\s+/g, '_')}.csv`;
            document.body.appendChild(a);
            a.click();
            window.URL.revokeObjectURL(url);
            document.body.removeChild(a);
        } catch (error) {
            alert('Export failed: ' + error.message);
        }
    },

    bindAllVessels() {
        const exportAggregatedBtn = document.getElementById('page-export-aggregated-btn');

        exportAggregatedBtn.addEventListener('click', async () => {
            try {
                const blob = await api.exportAggregatedCSV();
                const url = window.URL.createObjectURL(blob);
                const a = document.createElement('a');
                a.href = url;
                a.download = 'aggregated_vessels.csv';
                document.body.appendChild(a);
                a.click();
                window.URL.revokeObjectURL(url);
                document.body.removeChild(a);
            } catch (error) {
                alert('Export failed: ' + error.message);
            }
        });
    },

    async loadAggregatedVessels() {
        try {
            const data = await api.getAggregatedVessels();

            document.getElementById('page-total-unique-vessels').textContent = data.total_unique_vessels;

            const container = document.getElementById('page-aggregated-vessels-content');

            if (data.vessels.length === 0) {
                container.innerHTML = '<p style="text-align: center; opacity: 0.7; padding: 3rem;">No vessels found</p>';
                return;
            }

            container.innerHTML = data.vessels.map(vessel => `
                <div class="aggregated-vessel-item">
                    <div class="vessel-info-row">
                        <div>
                            <strong style="font-size: 1.1rem;">${vessel.mmsi || 'N/A'}</strong>
                            ${vessel.imo ? `<span style="opacity: 0.7; margin-left: 0.5rem;">IMO: ${vessel.imo}</span>` : ''}
                        </div>
                        <div style="display: flex; gap: 1rem; flex-wrap: wrap; align-items: center;">
                            ${vessel.name ? `<span><strong>Name:</strong> ${vessel.name}</span>` : ''}
                            ${vessel.flag ? `<span><strong>Flag:</strong> ${vessel.flag}</span>` : ''}
                        </div>
                        <div class="list-count-badge">
                            ${vessel.list_count} ${vessel.list_count === 1 ? 'List' : 'Lists'}
                        </div>
                    </div>
                    <div class="vessel-lists-badges">
                        ${vessel.lists.map(list => `
                            <span class="search-result-badge" style="background-color: ${list.list_color}33;">
                                <span class="list-badge" style="background-color: ${list.list_color}"></span>
                                ${list.list_name}
                            </span>
                        `).join('')}
                    </div>
                </div>
            `).join('');

        } catch (error) {
            console.error('Failed to load aggregated vessels:', error);
            alert('Failed to load vessels: ' + error.message);
        }
    },

    init() {
        this.loadStats();
        this.bindAdvancedSearch();
        this.bindAllVessels();
    }
};
